### ISSUE TITLE
Performance: Consider using Web Workers for more expensive calculations

### ISSUE THREAD

**Issue by miina** [GoogleForCreators/web-stories-wp#8253]
## Feature Description
[Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) allow running scripts in background threads. There are some limitations, for example, Web Workers can't access the DOM. However, when passing the image information to the Web Workers, some processing could still be done.

This would be useful especially for more "expensive" logic. Potential areas for using the Web Workers could be:
- `getPrepublishErrors` -- based on testing, running this could take 1+ seconds if there are many elements on the current page. EDIT: The pre-publish panel has already been improved meanwhile and this function doesn't exist anymore.
- Color contrast calculations done with the image generated by `html-to-image`. Even though the image will still need to be generated in the main thread (since Web Workers can't access DOM), all the following contrast calculation/processing can be done by Web Workers.
- `ColorThief` and average color for images -- not sure if this is possible, due to the DOM restriction.

Note: Web Workers should be able to fetch an image and get a blob from it so perhaps there are some other media-related optimisations possible using the Web Workers.

Some discovery is needed for detecting places for using the Web Workers.

## Alternatives Considered
Note -- if the function can be improved and there's an underlying issue causing performance problems, we need to fix that instead of using Web Workers.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Implementation Brief
Using `workerize` / `workerize-loader` might be a simple way to use Web Workers, then we should be able to do something like `import getPrepublishErrors from 'workerize!./getPrepublishErrors'`.

**Comment by swissspidy** [comment]
> * `getPrepublishErrors` -- based on testing, running this could take 1+ seconds if there are many elements on the current page.

Is this still an issue with the revamped checklist? Asking because there's no such function (anymore?).

I've mentioned this in the past for the old checklist (see #5768 and https://github.com/google/web-stories-wp/pull/5739#issuecomment-749135131) but I've always found the way errors are gathered to be very inefficient. For example there's no need to check for a missing story title over and over again when the title has never changed.
Instead, the error list could be implemented as side effects, only updating state when relevant changes were made (change story title -> run story title check -> maybe update checklist state).

Just moving this expensive calculation to a separate thread doesn't really solve the underlying issue.

> (since Web Workers can't access DOM)

https://github.com/ampproject/worker-dom might be useful perhaps

**Comment by maxyinger** [comment]
> Is this still an issue with the revamped checklist? Asking because there's no such function (anymore?).

Nope. Shouldn't be an issue anymore in PPCv2. We set up each check to be responsible for rendering itself in a self contained component. So now we don't need side effects or anything we can just utilize the `useContextSelector` built into use story and do something like this to only update on story title changes:

```
function SomeCheck() {
  const title = useStory((story) => story.title);
  return shouldThisCheckRender(title) ? <ThisCheck /> : null;
}
```

We may need to do a small pass making all the `useStory` calls more specific, but I think this mitigates a lot of the troubles we were seeing with `getPrepublishErrors` as well as being better suited to handle the visual/feature variability between cards

**Comment by miina** [comment]
Great to know about PPCv2 🎉 

> Just moving this expensive calculation to a separate thread doesn't really solve the underlying issue.

Yep, definitely agreed that if there's a problem, that should be fixed instead of shipping it away. Some discovery is needed to determine more good use cases for Web Workers. I've only looked into the initial render so far, most of these cases would likely be related to user interactions instead.

**Comment by swissspidy** [comment]
See also #10133 

tl;dr we now have precedence for using workers in the editor, but the DX could be improved.
